src/components/date-time-input/date-time-input.ts

```diff
@@ -0,0 +1,652 @@
+      errors = DateTimeUtil.validateMinMax(
+        this.value,
+        this.minValue!,
+        this.maxValue!,
+        this.hasTimeParts,
+        this.hasDateParts
+      );
+
+      if (Object.keys(errors).length > 0) {
+        this.invalid = true;
+      } else {
+        this.invalid = false;
+      }
+    }
+
+    return errors;
+  }
+
+  private get hasDateParts(): boolean {
+    const parts =
+      this._inputDateParts ||
+      DateTimeUtil.parseDateTimeFormat(this.inputFormat);
+
+    return parts.some(
+      (p) =>
+        p.type === DateParts.Date ||
+        p.type === DateParts.Month ||
+        p.type === DateParts.Year
+    );
+  }
+
+  private get hasTimeParts(): boolean {
+    const parts =
+      this._inputDateParts ||
+      DateTimeUtil.parseDateTimeFormat(this.inputFormat);
+    return parts.some(
+      (p) =>
+        p.type === DateParts.Hours ||
+        p.type === DateParts.Minutes ||
+        p.type === DateParts.Seconds
+    );
+  }
+
+  private get targetDatePart(): DatePart | undefined {
+    let result;
+
+    if (this.hasFocus) {
+      const partType = this._inputDateParts.find(
+        (p) =>
+          p.start <= this.inputSelection.start &&
+          this.inputSelection.start <= p.end &&
+          p.type !== DateParts.Literal
+      )?.type as string as DatePart;
+
+      if (partType) {
+        result = partType;
+      }
+    } else {
+      if (this._inputDateParts.some((p) => p.type === DateParts.Date)) {
+        result = DatePart.Date;
+      } else if (this._inputDateParts.some((p) => p.type === DateParts.Hours)) {
+        result = DatePart.Hours;
+      } else {
+        result = this._inputDateParts[0].type as string as DatePart;
+      }
+    }
+
+    return result;
+  }
+
+  private get datePartDeltas(): DatePartDeltas {
+    return Object.assign({}, this._datePartDeltas, this.spinDelta);
+  }
+
+  constructor() {
+    super();
+    this.addEventListener('wheel', this.onWheel);
+  }
+
+  public override connectedCallback() {
+    super.connectedCallback();
+    this.updateDefaultMask();
+    this.setMask(this.inputFormat);
+    if (this.value) {
+      this.updateMask();
+    }
+  }
+
+  /** Checks for validity of the control and shows the browser message if it's invalid. */
+  public reportValidity() {
+    const state = this._value
+      ? Object.keys(this.validate()!).length === 0
+      : this.input.reportValidity();
+
+    this.invalid = !state;
+    return state;
+  }
+
+  /** Check for validity of the control */
+  public checkValidity() {
+    if (this.disabled) {
+      return this.input.checkValidity();
+    }
+
+    if (!this._value) {
+      return !this.required;
+    }
+
+    return (
+      this.input.checkValidity() && Object.keys(this.validate()!).length === 0
+    );
+  }
+
+  public stepUp(datePart?: DatePart, delta?: number): void {
+    const targetPart = datePart || this.targetDatePart;
+
+    if (!targetPart || this.readonly) {
+      return;
+    }
```

Comment:
Not sure about preventing an API call if the input is read-only.
`stepUp/Down` should work in disabled/read-only state as it works with the default input.

@damyanpetev 
Your thoughts on this?

Reply:
Agreed, it's public API that should not be limited. Seems like this is due to the public method being used directly in event handling. The read only check should be moved there instead.